### PR TITLE
Add tag filtering

### LINKED ISSUE #36: Add slugified version of division, objectives, and types fields for CSS
CSS will expect slugified version of fields of known types. If those types change or types are added, the SCSS will need to be updated appropriately.

### PR DESCRIPTION
**Pull request by chriswhong**
This PR:
- adds dynamic tags chips to the sidebar in `Idea`
- adds dynamic tag chips to the sidebar in `Ideas`
- allows the user to click on a Tag Chip to filter the Ideas
- manages the URL query params for both categories (strategic objectives) and tags

`Ideas.js` is cleaned up a bit by making a generic `getChips()` function that works with both categories and tags, and by making a generic `changeFilter()` function that handles chip click events.

Closes #36

### CODE REVIEW

**Comment by chriswhong** [comment]
TODO: The tag Chips in the sidebar are orange, need to figure out the right way to make them gray via foundation @andycochran 

nyc_planning_labs

File: src/Ideas.js [hunk 1]
```diff
@@ -8,91 +8,116 @@ import './Ideas.css';
...
     const { ideas } = this.props;
-    const { categories, search } = this.state;
+    const { search } = this.state;
 
     // markup and event bindings for categories
-    const getObjectives = (objectives, header) => objectives
+    const getChips = (values, type, header) => values
       .map((d) => {
-        const disabled = (header && categories.indexOf(d) < 0) ? 'hollow' : '';
+        const disabled = (header && this.state[type].indexOf(d) < 0) ? 'hollow' : '';
```
Comment (allthesignals) [inline src/Ideas.js:115]: I think this function is a good candidate for refactoring. It's seems too intertwined with view logic and business logic. 

Maybe it should become its own component? 

Probably should table for later but curious to hear your thoughts.

File: src/Ideas.js [hunk 2]
```diff
@@ -8,91 +8,116 @@ import './Ideas.css';
...
-    const allAreSelected = checkIfAllSelected(categories, defaultSelection);
+    const paramChunks = [];
 
-    history.push({
-      search: allAreSelected ? '' : `?categories=${categories.join(',')}`,
-    });
+    if (!checkIfAllSelected(categories, allValues.categories)) paramChunks.push(`categories=${categories.join(',')}`);
+    if (!checkIfAllSelected(tags, allValues.tags)) paramChunks.push(`tags=${tags.join(',')}`);
+
+    const search = (paramChunks.length === 0) ? '' : `?${paramChunks.join('&')}`;
```
Comment (allthesignals) [inline src/Ideas.js]: This method (updateParams) is a little hard to parse, I think because there are too many ternaries and one-line conditionals. This may be a good candidate for refactoring. Considering this is a small project, it's probably not worth agonizing over too much, but I think it could be helpful to think about for later projects.

That being said, since React Router 4 did away with an opinionated approach to managing query parameters, we have room to find a good pattern for our purposes. I'd argue that the best approach would be to have 2-way data-binding between the query parameters and the state. Then, we are keeping the routing logic away from data munging and conditional logic. Updates to QPs change state, and updates to state change QPs accordingly.

File: src/Ideas.js [hunk 3]
```diff
@@ -8,91 +8,116 @@ import './Ideas.css';
 
 slug.defaults.mode = 'rfc3986';
 
-const defaultSelection = ['Economic Development', 'Data and Expertise', 'Resiliency and Sustainability', 'Neighborhood Improvement', 'Housing'];
+const allCategories = ['Economic Development', 'Data and Expertise', 'Resiliency and Sustainability', 'Neighborhood Improvement', 'Housing'];
 
-function checkIfAllSelected(categories, all) {
+const allValues = {};
+
+const checkIfAllSelected = (categories, all) => {
```
Comment (allthesignals) [inline src/Ideas.js]: This can be rewritten as:
```
const checkIfAllSelected = (categories, all) =>
  all.every(current => categories.indexOf(current) > 0);
```

File: src/Ideas.js [hunk 4]
```diff
@@ -8,91 +8,116 @@ import './Ideas.css';
...
+      currentSelected = removeItem(currentSelected, value);
     } else {
-      categories.push(clickedCategory);
+      currentSelected.push(value);
     }
 
-    if (categories.length === 0) categories = defaultSelection;
+    if (currentSelected.length === 0) currentSelected = all;
+
+    this.state[type] = currentSelected;
```
Comment (allthesignals) [inline src/Ideas.js]: Browser warning: `Do not mutate state directly. Use setState().`

Not sure how it affects things in this particular case (it's working!!), but I think it's a best practice.

File: src/Ideas.js [hunk 5]
```diff
@@ -8,91 +8,116 @@ import './Ideas.css';
...
     super();
-    const { location } = props;
+    const { location, ideas } = props;
     const query = new URLSearchParams(location.search);
-    const value = query.get('categories') || '';
+    const queryCategories = query.get('categories') || '';
+    const queryTags = query.get('tags') || '';
+
+    allValues.categories = allCategories;
+    allValues.tags = unique(ideas.reduce((a, b) => a.concat(b.tags), []).filter(Boolean));
```
Comment (allthesignals) [inline src/Ideas.js]: I know this is from one of my old commits, but I think with function argument names, we should go wild with how they're named (in terms of length). No need to use a,b,c - the minifier in the build step does that for us.

File: src/Ideas.js [hunk 6]
```diff
@@ -8,91 +8,116 @@ import './Ideas.css';
...
   all.forEach((d) => {
     if (categories.indexOf(d) < 0) allSelected = false;
   });
   return allSelected;
-}
+};
 
 // utility helpers for arrays
-function removeItem(array, value) {
+const removeItem = (array, value) => {
```
Comment (allthesignals) [inline src/Ideas.js]: Question: why does this need property initializer syntax here? I thought the main benefit of it was to bind `this` of the outer scope (or respective class).

**Review by allthesignals — CHANGES_REQUESTED** [review]
Not really "requesting changes", just had one or two things that might need fixing, depending on what you decide. I'm happy to address those things, but I wanted to hear your thoughts first. Otherwise, this works beautifully!

**Comment by andycochran** [comment]
I've updated the styles for 'filter by tag'. ;) https://github.com/NYCPlanning/labs-ideas/commit/a5e164da5ab3f10db2a5d80ee249a90e42ebbf53

File: src/Ideas.js (hunk 3 above)
Comment (chriswhong) [reply src/Ideas.js]: Will fix

File: src/Ideas.js (hunk 6 above)
Comment (chriswhong) [reply src/Ideas.js]: Can't stand behind it, not sure the value of setting a property initializer or using `function`

File: src/Ideas.js (hunk 2 above)
Comment (chriswhong) [reply src/Ideas.js]: leaving for now, maybe a package can help us with the state <-> query params thing later

File: src/Ideas.js (hunk 1 above)
Comment (chriswhong) [reply src/Ideas.js:115]: Yes, I was close to making this its own component, agree to table for now
